velox/expression/CastExpr.cpp

```diff
@@ -241,6 +241,45 @@ void applyIntToDecimalCastKernel(
   });
 }
 
+template <
+    typename TInput,
+    typename TOutput,
+    typename = typename std::enable_if_t<std::is_floating_point_v<TInput>>>
+void applyDoubleToDecimal(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& castResult) {
+  auto sourceVector = input.as<SimpleVector<TInput>>();
+  auto castResultRawBuffer =
+      castResult->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+  context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
+    const auto& sourceValue = sourceVector->valueAt(row);
+    if (!std::isfinite(sourceValue)) {
+      castResult->setNull(row, true);
+      return;
+    }
+
+    auto toValue = sourceValue *
+        static_cast<TInput>(DecimalUtil::kPowersOfTen[toPrecisionScale.second]);
+
+    if (toValue >= 0) {
+      toValue = std::floor(toValue + 0.5);
+    } else {
+      toValue = std::ceil(toValue - 0.5);
+    }
+
+    if (toValue <= static_cast<TInput>(std::numeric_limits<TOutput>::min()) ||
+        toValue >= static_cast<TInput>(std::numeric_limits<TOutput>::max())) {
```

Comment:
If value beyond the  range, it will throw exception in Presto.
```
presto:default> select cast((cast ('999999999999999999999.0' as double)) as decimal(18, 2));
Query 20230724_153614_00011_c9bdk failed: Cannot cast DOUBLE '1.0E21' to DECIMAL(18, 2)

presto:default> select cast((cast ('999999999999999999999.0' as double)) as decimal(38, 2));
           _col0
---------------------------
 1000000000000000000000.00
(1 row)
```